text/tls-and-alpn.rst

```diff
@@ -0,0 +1,380 @@
+TLS and ALPN
+============
+
+This RFC proposes to change the transport of EdgeDB frontend connections to TLS,
+and use ALPN for multiplexing protocol selection.
+
+
+Motivation
+----------
+
+The Transport Layer Security (TLS) Protocol is widely used for providing privacy
+and data integrity between two communicating applications [1]_. EdgeDB as a
+database should naturally support TLS as the transport layer protocol.
+
+With the help of EdgeDB tooling, it will be easy to setup TLS certificates for
+both development and production environments. Given also that the benchmark
+result of EdgeDB throughput overhead comparing TLS to vanilla TCP is about
+5% ~ 15%, it is also proposed to communicate in TLS by default.
+
+.. image:: ./EdgeDB-TLS-Benchmark.png
+
+At last, with TLS enabled by default, it is possible to leverage the TLS
+Application-Layer Protocol Negotiation (ALPN) Extension [2]_ for secure and
+reliable protocol selection on top of the TLS transport, allowing EdgeDB to
+multiplex different frontend protocols like the binary protocol, EdgeQL over
+HTTP or GraphQL over HTTP (and maybe more in the futre) on the same port as
+recommended in Pre-RFC 1910 [3]_.
+
+
+Overview
+--------
+
+EdgeDB server CLI (``edb server``) will take a few more parameters to configure
+a proper TLS server::
+
+    --tls-certfile
+    --tls-keyfile
+    --tls-private-key-password
+    --tls-protocol-min
+    --tls-protocol-max
+    --tls-ciphers
+
+The end user is not usually supposed to interact with this API directly, so we
+need to bring this to a higher user-level interface with convenience in mind.
+We shall discuss about that later in this RFC.
+
+The EdgeDB TLS server will by default advertise all protocols supported by the
+server using the ALPN Extension during TLS handshakes. However the way we
+advertise them is still pending for discussion. Protocols we have now:
+
+* The EdgeDB binary protocol
+
+* HTTP protocol, on top of which we have:
+
+  * System API starting with ``/server```
+
+  * Extensions that may differ from databases, starting with ``/db``:
+
+    * EdgeQL over HTTP
+
+    * GraphQL over HTTP
+
+    * "Notebook" - parameterless EdgeQL in HTTP-posted JSON
+
+All EdgeDB client bindings should use TLS to talk to an EdgeDB server.
+If possible, the client should also join the ALPN negotiation to choose its
+preferred protocol for each connection. The ALPN support in target languages:
+
+* Python [4]_: ``set_alpn_protocols()`` and ``selected_alpn_protocol()``
+* Go [5]_: ``SupportedProtos`` and ``NegotiatedProtocol``
+* Node.js [6]_: ``ALPNProtocols`` and ``alpnProtocol``
+
+If a specific protocol was successfully negotiated, that protocol will be used
+as the exclusive one throughout the lifetime of the connection. Or else if the
+ALPN agreement was not made, depending on the discussion below we will fallback
+to either the EdgeDB binary protocol or the current HTTP protocol.
```

Comment:
Don't see why we wouldn't make ALPN required.

We may try to detect clear-text connections for some cases. But maybe even this is not needed.

Reply:
> Don't see why we wouldn't make ALPN required.

+1.

Reply:
Sorry about the ambiguous wording! I think it also works for me to raise an error if a 3rd-party client fails to provide a valid ALPN protocol, if that's what you meant please?

Reply:
Yes.